### ISSUE TITLE
RowCountPrecondition fails on tables containing more than Integer.MAX_VALUE rows

### ISSUE THREAD

**Issue by kzander91** [liquibase/liquibase#1549]
## Environment
**Liquibase Version**: 4.1.0

**Liquibase Integration & Version**: CLI, maven

**Liquibase Extension(s) & Version**: none

**Database Vendor & Version**: Oracle Database 18c Enterprise Edition Release 18.0.0.0.0 - Production

**Operating System Type & Version**: SUSE Linux Enterprise Server 11 SP4

## Description
The `RowCountPrecondition` (and its subclass `TableIsEmptyPrecondition`) deal with the expected and actual row counts using `int` and `Integer`, causing change sets like this one

```xml
<changeSet ...>
 <preConditions>
  <tableIsEmpty tableName="LARGE_TABLE"/>
 </preConditions>
</changeSet>
```

to fail if the table in question contains more than 2147483647 (`Integer.MAX_VALUE`) rows.
Additionally, _counting_ the number of rows in a table just to check if it is empty is highly inefficient.

## Steps To Reproduce
Create a table and insert more than 2147483647 rows, then check its row count using the `<tableIsEmpty/>` or `<rowCount/>` preconditions. Their implementations count the rows using `JdbcExecutor.queryForInt(..)`, which eventually calls `ResultSet.getInt()`.

## Actual Behavior
The following exception is thrown:

```
Caused by: java.sql.SQLException: Numeric Overflow
	at oracle.jdbc.driver.NumberCommonAccessor.throwOverflow(NumberCommonAccessor.java:4139)
	at oracle.jdbc.driver.NumberCommonAccessor.getInt(NumberCommonAccessor.java:125)
	at oracle.jdbc.driver.GeneratedStatement.getInt(GeneratedStatement.java:199)
	at oracle.jdbc.driver.GeneratedScrollableResultSet.getInt(GeneratedScrollableResultSet.java:246)
	at oracle.ucp.jdbc.proxy.oracle$1ucp$1jdbc$1proxy$1oracle$1ResultSetProxy$2oracle$1jdbc$1internal$1OracleResultSet$$$Proxy.getInt(Unknown Source)
	at liquibase.executor.jvm.SingleColumnRowMapper.getColumnValue(SingleColumnRowMapper.java:130)
```

## Expected/Desired Behavior
These preconditions should work normally on tables with more than 2147483647 rows.
I suppose the most straight forward fix would be to switch the types used to `long`/`Long` and use `queryForLong` instead.

**Comment by molivasdat** [comment]
Hi @kzander91 Thanks for writing up this issue.  We will add it to the list of issues to process.  I agree that counting all the rows in a table is very inefficient to see if the table is empty.

**Comment by tati-qalified** [comment]
Hello @kzander91, sorry for the late response, we're going through our backlog of issues and trying to catch up.
Both things you're saying make sense; it's not efficient to count all rows to see if a table is empty, and Liquibase should support tables with more rows than the max integer value.

A good solution for the efficiency would be to select the first row of the data object and check if it's empty or not.
For the amount of rows, I agree with what you've suggested.

Would you (or anyone else who is reading this) be willing to submit a PR?
Thank you!
Tatiana

**Comment by kzander91** [comment]
Hi @tati-qalified, I just opened #4725 to change the type to `long`.
I considered implementing the optimization of `TableIsEmptyPrecondition` by fetching the table's first row, but it looks like `Database` doesn't have support yet for generating limit clauses (`select top ? from ...`, `select ... from ... limit ?`, `select ... from ... fetch first ? rows only`, etc.). I found some rather convoluted logic that does that here:
https://github.com/liquibase/liquibase/blob/d6458d45bb55d85a82dc95f223152b8366b49a75/liquibase-standard/src/main/java/liquibase/sqlgenerator/core/SelectFromDatabaseChangeLogGenerator.java#L40-L90

This should probably be a feature of the `Database` abstraction, such that changes and preconditions can reuse it. I didn't feel comfortable doing such a large refactoring as a first contribution.
